lib/content-patcher/patchers/class-shortcodepullquotepatcher.php

```diff
@@ -0,0 +1,103 @@
+<?php
+/**
+ * Patcher for the [pullquote][/pullquote] elements.
+ *
+ * @package Newspack
+ */
+
+namespace NewspackContentConverter\ContentPatcher\Patchers;
+
+use NewspackContentConverter\ContentPatcher\Patchers\PatcherInterface;
+use NewspackContentConverter\ContentPatcher\Patchers\PatcherAbstract;
+use NewspackContentConverter\ContentPatcher\ElementManipulators\SquareBracketsElementManipulator;
+use NewspackContentConverter\ContentPatcher\ElementManipulators\WpBlockManipulator;
+use NewspackContentConverter\ContentPatcher\ElementManipulators\HtmlElementManipulator;
+
+/**
+ * Patcher class for the [pullquote][/pullquote] elements.
+ *
+ * @package NewspackContentConverter\ContentPatcher\Patchers
+ */
+class ShortcodePullquotePatcher extends PatcherAbstract implements PatcherInterface {
+
+	/**
+	 * SquareBracketsElementManipulator service.
+	 *
+	 * @var SquareBracketsElementManipulator
+	 */
+	private $square_brackets_element_manipulator;
+
+	/**
+	 * WpBlockManipulator service.
+	 *
+	 * @var WpBlockManipulator
+	 */
+	private $wp_block_manipulator;
+
+	/**
+	 * VideoPatcher constructor.
+	 */
+	public function __construct() {
+		$this->square_brackets_element_manipulator = new SquareBracketsElementManipulator();
+		$this->wp_block_manipulator                = new WpBlockManipulator();
+	}
+
+	/**
+	 * See the \NewspackContentConverter\ContentPatcher\Patchers\PatcherInterface::patch_blocks_contents for description.
+	 *
+	 * @param string $source_html   HTML source, original content being converted.
+	 * @param string $source_blocks Block content as result of Gutenberg "conversion to blocks".
+	 *
+	 * @return string|false
+	 */
+	public function patch_blocks_contents( $source_html, $source_blocks ) {
+		$matches_blocks = $this->wp_block_manipulator->match_wp_block( 'wp:shortcode', $source_blocks );
+		if ( ! $matches_blocks ) {
+			return $source_blocks;
+		}
+
+		foreach ( $matches_blocks[0] as $matched_block ) {
+			$block = $matched_block[0];
+			if ( false === strpos( $block, '[/pullquote]' ) ) {
+				continue;
+			}
+
+			$converted_block = $this->convert_shortcode_block_to_pullquote( $block );
+			$source_blocks   = str_replace( $block, $converted_block, $source_blocks );
+		}
+
+		return $source_blocks;
+	}
+
+	/**
+	 * Convert a shortcode block with the pullquote shortcode into a pullquote block.
+	 *
+	 * @param string $block Raw block content.
+	 * @return string New block content.
+	 */
+	protected function convert_shortcode_block_to_pullquote( $block ) {
+		// Remove newlines because they confuse the matchers.
+		$block = str_replace( "\n", '', $block );
```

Comment:
I've tried commenting out this line, and the tests still work fine. Do you maybe have an example of where removing the "\n" should make a difference?

Reply:
In the future, we should probably fix in the main regex that does the matching, but it's not a super straightforward fix (I believe some existing unit tests break). Comment out the above line again and change `get_unpatched_block` in the unit test to:

```
<!-- wp:shortcode -->
[pullquote 
	author="Mary Filardo, president of the 21st Century School Fund" 
	description="" 
	style="new-pullquote"]“It’s just one more nail in the coffin of small towns that are already struggling. The county hospital closed, and the mom-and-pop shops are gone because Walmart opened. When you lose the schools, you lose the community.” [/pullquote]
<!-- /wp:shortcode -->
```

You will see that nothing really gets matched:
```
1) TestPullquoteShortcodePatcher::test_patch_pullquote_shortcode
Failed asserting that two strings are identical.
--- Expected
+++ Actual
@@ @@
 '<!-- wp:pullquote {"align":"left"} -->
-<figure class="wp-block-pullquote alignleft"><blockquote><p>“It’s just one more nail in the coffin of small towns that are already struggling. The county hospital closed, and the mom-and-pop shops are gone because Walmart opened. When you lose the schools, you lose the community.”</p><cite>Mary Filardo, president of the 21st Century School Fund</cite></blockquote></figure>
+<figure class="wp-block-pullquote alignleft"><blockquote><p></p></blockquote></figure>
```

Reply:
Thanks for sharing this. Got it.
You're right, it's pretty easy making this test pass without stripping the "\n:" by adding multiline support to these two regexes (adding the `s` modifier to the regex):
- https://github.com/Automattic/newspack-content-converter/blob/master/lib/content-patcher/elementManipulators/class-squarebracketselementmanipulator.php#L29
- https://github.com/Automattic/newspack-content-converter/blob/master/lib/content-patcher/elementManipulators/class-squarebracketselementmanipulator.php#L44

but then other four tests fail, which all of them test whether _the patchers work with invalid source_, an example:
- https://github.com/Automattic/newspack-content-converter/blob/master/tests/fixtures/unit/content-patcher/patchers/class-dataprovideraudiopatcher.php#L368
(if multiline is enabled, then for this example test it will match the opening of the 2nd audio tag with closing of the 3rd one, which would not be correct).

Let me ask you for an opinion/advice. It's very possible that I was overly paranoid when creating tests _with invalid source_ to make sure that the patchers would work even with those. But I wonder, is that even necessary? Should we remove these tests with "invalid sources"? I could create another PR, remove these tests, and then also come back to this code of yours, and later on remove the stripping of the "\n"?

Reply:
Yeah I think we can remove the tests with invalid sources. Since we're converting existing published content, the shortcodes should (in general) be valid. Multiline shortcodes will be much more common than invalid shortcodes. In theory, I wouldn't expect any invalid shortcodes, as they would have caught those when publishing the post in the first place. :)

Reply:
Ok, thanks. Let's merge this one, and I'll create a new PR referencing this, and cleaning up a bit when we get to it :thumb_up: